### ISSUE TITLE
Discussion: initial work on v2

### ISSUE THREAD

**Issue by rtfb** [russross/blackfriday#218]
I have rebased my v2 experiments on top the latest master and pushed them to v2 branch for eyeballing.

Here's what I was doing and where I stand:
- All constants renamed from C_STYLE to IdiomaticGo (and plain ints to typed
  constants). The rename was mechanical so far, I didn't try to take a chance
  and look for opportunities to eliminate naming inconsistencies; but we
  probably should do that while we're breaking the compatibility. Suggestions
  welcome.
- Renderer interface overhaul. Eliminated callbacks from renderer methods.
  #120 suggested changing `*bytes.Buffer` to `io.Writer`, but I tried another
  approach: leaving all writing business to the renderer internals.
- #189 is probably fixed as side effect: no more callbacks, no more dead
  loops.
- When table of contents is enabled, it still behaves the old way: backups
  all the document in a temp buffer, truncates it, writes out TOC, writes back
  temp buffer. I don't see an easy way around it.

The next big thing I'm going to try is constructing an actual AST before rendering. This should allow to clean up our internal code substantially in some places and further simplify the renderer interface (and properly fix the TOC problem).

Comments welcome.

**Comment by russross** [comment]
I haven't had a chance to look at it yet, but from your description I can say that this sounds like the right direction to me. I especially like the idea of moving to an AST where users can manipulate it freely and still use the normal HTML renderer when the are finished making changes.

The only thing I would add is this: don't feel like you need to wait for my approval when you are ready to commit to these large changes. When you think it is ready, you should be the one to make the call.

Thanks again for all of your work on this project!

**Comment by egonelbre** [comment]
Some tips. Introduce a line-reader into the code. See https://github.com/loov/mark/blob/master/reader.go

This allows to write: 
* [main loop](https://github.com/loov/mark/blob/master/parse.go#L90)
* [fenced code](https://github.com/loov/mark/blob/master/parse.go#L267)
* [quoted blocks](https://github.com/loov/mark/blob/master/parse.go#L134)

*Of course, adjust reader to your own needs. e.g. [something smaller](https://github.com/egonelbre/mmark/blob/50dfad50e1706d9bd04c9bd35c0a369baf424449/markdown.go#L1002)*

**Comment by client9** [comment]
Hi team,

This is a fantastic project.  

When using the API I bumped into a few oddities, that I'd love to see fixed up in a v2.  I'm happy to help or work on any of these idea.

##### Would like to know the position in the original markdown text.

The API has no idea where in the source text it is, and it's not really possible to compute it since we don't know how big the raw input token was.  An AST would help here since it would likely have a position of the token found (either in bytes or line/column)

##### The output buffer is used as a stack in the main code

I wasn't writing to  `out *bytes.Buffer` at all for my project since I wasn't rendering but really scanning the text for things of interest.  It turns out a few places in the code that uses the output buffer to make decisions (if this the first character? is the last character an `!`, etc)

In particular the `NormalText` func seems to write the buffer, and then in-line text is seems to pop this off and redo it.  No doubt for good reason, but the output buffer as stack was confusing.

And to keep things honest...

##### API doesnt need a `bytes.Buffer` or `io.Writer` at all.

Currently

```go
func (mt *markdownParser) CodeSpan(out *bytes.Buffer, text []byte) {
   out.WriteString("<code>")
   out.Write(text)
   out.WriteString("</code>")
}
```

the caller can do anything they want with their own object/struct, eg.

```go
type markdownParser struct {
     out bytes.Buffer  <-- or whatever, or nothing at all
}

// no more out buffer in api
func (mt *markdownParser) CodeSpan(text []byte) {
   mt.out.WriteString("<code>")
   mt.out.Write(text)
   mt.out.WriteString("</code>")
}
```

Eliminating this makes the blackfriday API cleaner and more flexible.   And prevents cheating

##### Might be able to re-use x/net/html interfaces

Since we are more or less making a quasi-HTML parser, I'd love to see the future API (or a low-level API) follow existing parsers or objects from  [x/net/html](https://godoc.org/golang.org/x/net/html).  (e.g. [Token](https://godoc.org/golang.org/x/net/html#Token), Tokenizer)

the much from the https://godoc.org/golang.org/x/net/html API  

##### Provide an Identity renderer for testing, learning, etc

This would be similar to @shurcooL https://github.com/shurcooL/markdownfmt but without the reformatting.  If the input was well-formed the output would be identical to the input.  This would be a short intro the API and probably excellent for testing.

Thanks again.  I hope you found the constructive, and again, I'm happy to help.

regards,

n

**Comment by client9** [comment]
ok as followup, I just looked at the [commonmark spec](http://spec.commonmark.org/0.23/)  The specs is way more complicated than I expected and allows way more variation than I expected (look at the many forms for <hr>!!).  The implementations are just as complicated and the JS one is highly regex based (boo).

I can live with the output buffer now :-)   I can generate a small XML/HTML output using the API, then use x/net/html to process the result.  It's a poor mans AST  but it's not performance critical for me, and it allows full customization without much work.  I'll post an example, as it may be a good way for people to work around some of the issues people have been having with customization.

**Comment by rtfb** [comment]
Hi all,

I've finally got a chance to clean up my stuff on AST and push it. `v2` branch
now has a new function `Parse`, which takes input and spits out an AST. The
AST can then be processed in whatever way the user fancies (here's an example
of rendering everything the default way, but overriding the `<hr>`:
https://gist.github.com/rtfb/08babeb67d37514f5322722797a6bd6a).

The old `Markdown` and `MarkdownCommon` funcs are certainly still available
for convenience.

Some things that have no tests are still missing (e.g. TOC, CompletePage and
maybe some other flags are not implemented yet).

An important drawback of this thing is that it slows down things quite
considerably, around 2-3x (full test suite for v1 runs ~20-25secs on my box,
but ~65-70secs on v2). A quick look at profiler data indicates that the new
more intensive work with memory is to blame. Hopefully I can streamline it
later.

Now when the full AST is exposed, the natural question arises: is there any
need to have the elaborate Renderer interface? I have added a `ForEachNode`
func that calls a callback for every node for easy traversal of the tree. Is
there a need and value in anything else?

A technical nitpick: newlines. I have tried to write the new code without
touching any tests at all (and almost succeeded, only two minor exceptions:
a32ecdb, 6fe811b). However, I have found that there's a lot of elaborate
details in the way the newlines are produced in the old renderer. Sometimes we
produce newlines, sometimes we don't, some things are separated by blank
lines, but most are not, and so on. Is there any value in that? I'd like to
simplify that a lot, e.g. not produce any newlines within blocks and separate
blocks by single newlines or a similar simple rule. Any objections or
insights?

Any other comments and thoughts welcome.

**Comment by shurcooL** [comment]
I left a quick comment on the gist you shared.

**Comment by rtfb** [comment]
Some late answers to the thread...

@egonelbre,

> Suggestion. Introduce a line-reader into the code

I'd like to. That would probably let us clean up the parser a little bit. But
that is internals-only and I'm giving priority to externally visible things
now to get v2 done. I'll probably give it a try some time later.

@client9,

> Would like to know the position in the original markdown text.

Me too. I'm planning to do that, but haven't rushed with it for the first round.

> The output buffer is used as a stack in the main code

Not anymore :-)

> API doesnt need a bytes.Buffer or io.Writer at all.

Agreed. And as I wrote above, I'm not even sure we heed any API for the
renderer at all, the tree itself will suffice.

> Might be able to re-use x/net/html interfaces

Hmm, that's an interesting thought. Markdown is quite line-based, though, so
it might turn out not as practical as it seems, but that's a guesswork, need
to try.

> Provide an Identity renderer for testing, learning, etc

That's a great idea! I'll give it a try.

**Comment by rtfb** [comment]
One more thing surfaced. I started working on the TOC extension and noticed that it's a flag to the HTML _renderer_. To me the most natural way to implement it seems to be a list node in the document tree (and then the renderer doesn't need to know nothing about it). So I propose moving this flag to extensions. Does that make sense?

**Comment by rtfb** [comment]
After my last push v2 branch is now functionally equivalent to v1, AFAICT.

**Comment by djui** [comment]
`blackfriday.Parse`'s signature does not return errors. What is the parsing strategy on malformatted markdown documents parsed? Just leaving the broken parts as pure text?

**Comment by djui** [comment]
I can't find a `switch node.Type == md.Document && entering == false`. Is that intended? Should it be inferred when the AST is entirely walked that the document is leaving?

**Comment by rtfb** [comment]
> blackfriday.Parse's signature does not return errors. What is the parsing strategy on malformatted markdown documents parsed? Just leaving the broken parts as pure text?

Correct, anything not recognized as known markdown is treated as plain text.

> I can't find a node.Type == md.Document && entering == false node. Is that intended? Should it be inferred when the AST is entirely walked that the document is leaving?

Do you mean that the traverser does not produce a call to the walker func with `node.Type == md.Document && entering == false` arguments? It should, that's a bug if it doesn't.

**Comment by djui** [comment]
> Do you mean that the traverser does not produce a call to the walker func with node.Type == md.Document && entering == false arguments? It should, that's a bug if it doesn't.
I can't see the `!entering part for the node type Document. I also see that not all nodes have a `!entering` part.

**Comment by rtfb** [comment]
> I can't see the `!entering` part for the node type Document. I also see that not all nodes have a `!entering` part.

The ones that do not have `if-entering-else` distinction are leaf nodes. They're visited only once, so there's no need to guard against that. The `Document` node type is the root of the tree, it should be visited twice, entering and leaving, but since the root node is only a container and does not render anything, I didn't care much about it and left it as no op.

**Comment by Ambrevar** [comment]
I've sent a PR #271 for v2 that adds the math extension.

**Comment by Ambrevar** [comment]
I really like the job you did for v2. I've got a few questions however.

* Is there a necessity to distinguish between inline and block? Can you provide an example?

* Is HTMLRenderer.w (HTMLWriter) used at all?

* Running gometalinter reports quite a lot of lint issues. Most notably:

	> html.go:318:1:warning: isSmartypantable is unused (deadcode)
	> test_helpers.go:52:1:warning: doTestsBlock is unused (deadcode)
	> test_helpers.go:88:1:warning: doLinkTestsInline is unused (deadcode)
	> test_helpers.go:103:1:warning: doSafeTestsInline is unused (deadcode)
	> test_helpers.go:138:1:warning: doTestsReference is unused (deadcode)
	> node.go:317:1:warning: dump is unused (deadcode)

* The source says

	> // HTMLRenderer is a type that implements the Renderer interface for HTML output.
	> //
	> // Do not create this directly, instead use the NewHTMLRenderer function.

	Which makes sense since the map needs init. But your above gist does not use it.

* We can set `Extensions` in `HTMLRendererParameters` which are not necesserily the same as the parser Extensions. Is this a design issue?

	``` go
	renderer := bf.NewHTMLRenderer(bf.HTMLRendererParameters{
		Extensions: bf.LaTeXMath,
	})
	withmath := bf.Markdown(md, renderer, bf.Options{Extensions: bf.CommonExtensions})
	```

**Comment by Ambrevar** [comment]
* Strange issue: when blackfriday is imported, the Go testing flags get appended to the CLI flags as seen when calling `flag.Usage()`. Example code:

```go

package main

import (
	"flag"
	"fmt"

	bf "gopkg.in/russross/blackfriday.v2"
)

func main() {
	var defaultRenderer bf.HTMLRenderer
	fmt.Println(defaultRenderer)
	flag.Usage()
}
```

* The autolink extension will split text nodes when there is no match. Would be better to keep the text as one node.

* I am not sure it's right to create the TOC from the parser: you've included HTML code, so this function is HTML-specific. Besides, `HTMLRenderer` has an unused `toc` member. LaTeX does not need any TOC parsing at all. I believe the current AST exposes the TOC to its simplest form for all the other renderers that needs a custom TOC rendering.

* I've PR'ed a `(*Node) deepCopy() *Node` function. I wonder if this would be useful enough to be exported. On a more general level, @russross mentioned that he likes that the user can manipulate the AST. Your Gist, however, shows another approach: it manipulates the walk. I believe both approaches are equivalently powerful, but AST modifications can be tricky and fragile. Shall we disallow AST modifications?

	* To reply to @rtfb question on whether to expose functions other than `Walk()`: If we keep AST modifications, we could export the functions `insertNode()` and `appendChild()` as `golang.org/x/net/html` does. What about adding a `RemoveChild()`?

* The `dump*` functions do not seem be used much. I've sent a PR adding to make `Node` implement the `Stringer` interface using the code of `dump`. It's more standard, more convenient, and has the benefit that it can be called from external packages which is helpful for debugging. Shall we remove the 'dump*' functions and move keep only `String()`?

**Comment by shurcooL** [comment]
* Strange issue: when blackfriday is imported, the Go testing flags get appended to the CLI flags as seen when calling `flag.Usage()`.

That's because of `test_helpers.go` file, I suspect it should be have a `_test.go` suffix so that it's only included when building package for tests. That file imports `testing` package, which is what adds flags. This is easy to fix.

**Comment by rtfb** [comment]
Thanks for your helpful feedback, @Ambrevar!

> I've sent a PR #272 for v2 that adds the math extension.

Thanks! I'm completely unfamiliar with MathJax, so it takes time to grok. I'll
try to comment on the PR as soon as I can.

> Is there a real necessity to distinguish between 'inline' and 'block'? Can you
> provide an example?

Which distinction between inlines and blocks do you have in mind? If you mean
the parsing in stages, that's the way Blackfriday has been written from the
early days.

If you mean the nodes, there is not much difference between them by the time the
AST is complete. The only difference that I remember is the `Node.content` field
that carries the block content from block parser to inline parser. After the
block node has been processed by inline parser, the content is moved to the
child nodes and `Node.content` is discarded.

> Is `HTMLRenderer.w` (HTMLWriter) used at all?

Not any more, fixed :-)

> Running gometalinter reports quite a lot of lint issues. Most notably:
>
>     html.go:318:1:warning: isSmartypantable is unused (deadcode)
>     test_helpers.go:52:1:warning: doTestsBlock is unused (deadcode)
>     test_helpers.go:88:1:warning: doLinkTestsInline is unused (deadcode)
>     test_helpers.go:103:1:warning: doSafeTestsInline is unused (deadcode)
>     test_helpers.go:138:1:warning: doTestsReference is unused (deadcode)
>     node.go:317:1:warning: dump is unused (deadcode)

Yeah, that doesn't surprise me, haven't had time to address lint issues yet.

> The source says
>
>     // HTMLRenderer is a type that implements the Renderer interface for HTML output.
>     //
>     // Do not create this directly, instead use the NewHTMLRenderer function.
>
> Which makes sense since the map needs init. But your above gist does not use
> it.

Gist fixed, and some supporting changes in the library were also necessary.
Thanks for the catch.

> We can set Extensions in HTMLRendererParameters which are not necesserily the
> same as the parser Extensions. Is this a design issue?
>
>     renderer := bf.NewHTMLRenderer(bf.HTMLRendererParameters{
>             Extensions: bf.LaTeXMath,
>     })
>     withmath := bf.Markdown(md, renderer, bf.Options{Extensions: bf.CommonExtensions})

I think that was a quick hack to allow the renderer to see the specified
Extensions and react where necessary. Needs to be fixed, but don't know how yet.
Added an issue for tracking: #277.

> The autolink extension will split text nodes when there is no match. Would be
> better to keep the text as one node.

Yup, seen that and have a TODO for it somewhere. But won't have time to work on
it the nearest future, so if anybody wants to deal with it, say so and go for
it.

> I am not sure it's right to create the TOC from the parser: you've included
> HTML code, so this function is HTML-specific. Besides, HTMLRenderer has an
> unused toc member. LaTeX does not need any TOC parsing at all. I believe the
> current AST exposes the TOC to its simplest form for all the other renderers
> that needs a custom TOC rendering.

Agreed. It would be better to move the `generateTOC()` call to the beginning of
`HTMLRenderer.Render()`. Or probably even better, drop the `TOC` flag and export
the `generateTOC()` instead of it, so that people could decorate their ASTs even
in the cases that we do not foresee.

> I've PR'ed a (*Node) deepCopy() *Node function. I wonder if this would be
> useful enough to be exported. On a more general level, @russross mentioned
> that he likes that the user can manipulate the AST. Your Gist, however, shows
> another approach: it manipulates the walk. I believe both approaches are
> equivalently powerful, but AST modifications can be tricky and fragile. Shall
> we disallow AST modifications?

The idea to disallow AST modifications doesn't sound right to me. The whole
point of AST is to give the caller more flexibility than we can foresee.

> To reply to @rtfb question on whether to expose functions other than `Walk()`:
> If we keep AST modifications, we could export the functions `insertNode()` and
> `appendChild()` as `golang.org/x/net/html` does. What about adding a
> `RemoveChild()`?

Sounds good.

> The `dump*` functions do not seem be used much. I've sent a PR adding to make
> Node implement the Stringer interface using the code of dump. It's more
> standard, more convenient, and has the benefit that it can be called from
> external packages which is helpful for debugging. Shall we remove the `dump*`
> functions and move keep only `String()`?

That one is being discussed under the PR.

Thanks again! :-)

**Comment by Ambrevar** [comment]
> Which distinction between inlines and blocks do you have in mind? If you mean
> the parsing in stages, that's the way Blackfriday has been written from the
> early days.

Yep, that's what I meant. Since blackfriday is currently being rewritten,
wouldn't it be worthwhile removing the block/inline separation? I believe it
would make the code much simpler, shorter, and a little faster. (Which is yet to
be proven of course.) I might miss something though.

> > Is HTMLRenderer.w (HTMLWriter) used at all?

> Not any more, fixed :-)

Isn't it actually a better idea to keep the buffer inside the structure? All the
calls use it, so it would save one parameter. It's a nice idea to add a 'buffer'
parameter to the constructor though.

> Yup, seen that and have a TODO for it somewhere.

I cannot find this TODO you mentioned.

> Agreed. It would be better to move the generateTOC() call to the beginning of
> HTMLRenderer.Render(). Or probably even better, drop the TOC flag and export
> the generateTOC() instead of it, so that people could decorate their ASTs even
> in the cases that we do not foresee.

Ack.
I've sent a PR fixing the TOC generation. Then I suggest we do as you said: move
the function to HTML, export it, remove the call and the flag.

**Comment by Ambrevar** [comment]
v1 has its blackfriday-tool. What about v2? What about a canonical converter in blackfriday/cmd?

**Comment by Ambrevar** [comment]
I'm in the process of writing the LaTeX extension for v2: so far so good. The current TOC generation conflicts with LaTeX though: as we discussed above, it needs to be moved to the HTML renderer.

**Comment by Ambrevar** [comment]
Same thing regarding footnotes: the code needs be moved to the HTML renderer as this is HTML specific.
For LaTeX, a simple `\footnote{foobar}` is enough. As such, the parser needs be changed so that the RenderingNode function can access the footnote text immediately from the current 'node'. Maybe add a dedicated Node type? The current implementation seems to be way too complicated.

**Comment by rtfb** [comment]
> I'm in the process of writing the LaTeX extension for v2: so far so good.

That's great news!

> The current TOC generation conflicts with LaTeX though: as we discussed above, it needs to be moved to the HTML renderer.

Yeah... I'm sorry it took so long to hear from me on this subject, but #273 is quite involving and I'm trying to think of some workarounds to avoid deep copying. I'll go dump my comments on it right away and when we settle with it, I can move TOC to html renderer.

**Comment by rtfb** [comment]
> v1 has its blackfriday-tool. What about v2? What about a canonical converter in blackfriday/cmd?

I was thinking to patch up blackfriday-tool itself to use v2 when it's done. And I'm in favor of moving blackfriday-tool to blackfriday/cmd. It's quite typical to do so in Go world.

@russross, what do you think?

**Comment by shurcooL** [comment]
> And I'm in favor of moving blackfriday-tool to blackfriday/cmd. It's quite typical to do so in Go world.

I think it should be `github.com/russross/blackfriday/cmd/blackfriday`, so that the binary name is `blackfriday` rather than `cmd`.

**Comment by russross** [comment]
blackfriday-tool was part of the main repo originally. I pulled it out because I guessed that most people were just interested in the API and I wanted to remove clutter (same reason I think the LaTeX output should probably be split off). Also, at the time the community had not settled on the library/ and library/cmd structure that is common for Go projects now.

Honestly, I don't have a strong opinion one way or the other. As long as the command-line tool stays out of the way for API users (which it probably would in a subdirectory) I think it is fine either way.

**Comment by rtfb** [comment]
@shurcooL, I did mean moving it to this repo: `github.com/russross/blackfriday-tool` -> `github.com/russross/blackfriday/cmd`. But this is not a strong opinion in any way. I don't think this question is of big importance for us at this point or near future.

**Comment by shurcooL** [comment]
I see.

Well, I'm fully in agreement about changing `blackfriday-tool` to use v2 whenever it's ready.

But let's hold off on moving it for now (and make a separate issue/proposal when it's time).

**Comment by Ambrevar** [comment]
It does not matter much whether it has it own repo or not, but for one thing: I think "blackfriday-tool" is a poor naming while the simple "blackfriday" would be much nicer.

**Comment by Ambrevar** [comment]
We've made quite some progress last month: TOC generation is fixed, most of the lint issues are gone, the walk is fixed, the aforementionned block issue is gone, etc.

But where are we now? Is there any development route? The wiki is empty. Or shall we just rely on the issue list?

What remains to be done, as far as I see it:

- Autolink needless breaking when there is no link: not a blocker but a bit ugly and possibly a performance killer. I can look into it.

- Use consistent line breaking rules: I've done that for the LaTeX renderer, I can push a similar PR for the HTML renderer. My rule of thumb is as follows: never break lines before, always after.

- Code hooks for math, possibly tables, and the rest: this is being discussed in #288. I think table and math are quite important. @rtfb suggested a nice design compatible with Common Mark.

- Commonmark validation: we should be quite close.

- Port the command line tool: this is easy. I can do it as well.

- Footnotes should not change the AST beyond their own node: this is a big blocker for other renderers (namely LaTeX). From a quick git-blame I think @rtfb wrote most of it. Do you have time to fix this or should I give it a go? This one might be a bit challenging and you certainly have more knowledge than me on the current implementation.

**Comment by Ambrevar** [comment]
Footnotes: @rtfb suggested we keep the list at the end of the AST and we add a reference to the nodes from the link structure. It works fine with my LaTeX renderer. 

The needless node splitting for Autolink and other has been fixed too.

**Comment by rtfb** [comment]
Regarding CommonMark, I've been thinking of it as a future project to tackle after v2 is finalized and released. I think the major improvement that we bring with v2 is API cleanup and untying people's hands for 3rd party development.

Whereas CommonMark is mostly orthogonal to that. It should only affect the internals, touch little or nothing on the API side (hope so!), so holding up the API changes while we're tinkering with the internals makes no sense.

**Comment by Ambrevar** [comment]
Is your intent with v2 to have a stable API and not necessarily stable parsing?
What do you think of offering a stable parsing guarantee?

**Comment by rtfb** [comment]
> Is your intent with v2 to have a stable API and not necessarily stable parsing?
> What do you think of offering a stable parsing guarantee?

Depends on what you mean by stable parsing guarantee. I am trying to hold on to stable feature set as much as possible, but I don't want to commit to any strong guarantees yet, since the spec is so much in flux and we have to apply some... creative interpretation to certain corner cases of input.

#289 is a good example of a gray area: supporting pluses instead of pipes seems to be such an innocent change, and it doesn't bring a new feature, only add some convenience to an existing one. Had we frozen the supported source format, we'd have to say no.

**Comment by Ambrevar** [comment]
Ack. It does not make any sense to guarantee parsing if there is no stable specs anywhere.

I strongly dislike where CommonMark is going with its over-complicated emphasis parsing rules (as I mentioned in #316).

Regarding #289: I think we should close this as it does not belong to BF, really. Besides, we cannot afford implementing every single nits if we want to avoid drifing toward  some bloated, unmaintainable library. Let's keep it simple and to the point:

- As close as possible to the "widely accepted standard" (more or less CommonMark).
- Keep the extensions we have, refrain on accepting anything new unless overwhelmingly essential.

**Comment by Ambrevar** [comment]
I have finished polishing the [LaTeX renderer](http://bitbucket.org/ambrevar/blackfriday-latex): everything is implemented, test covereage is of 75%+, it looks pretty good to me.

Reviews are welcome.
